### PR TITLE
fix(twap/upgrades): geometric twap accumulator initialization; testnet bug reproduction

### PR DESCRIPTION
**Pull request by p0mvn**
Closes: #XXX

## What is the purpose of the change

In https://github.com/osmosis-labs/osmosis/issues/3542, it was concluded that we do not need to perform the migrations to initialize the geometric twap accumulator.

However, the testnet halted due to the geometric twap accumulator ending up uninitialized.

The reason why our e2e test did not catch this problem was because we neve swapped against a pool that was created pre-upgrade after the upgrade itself.

This PR reproduced the testnet halt bug and fixed it by introducing twap record migrations that were originally spearheaded  in: https://github.com/osmosis-labs/osmosis/pull/3729

## Brief Changelog

- reproduce testnet halt
- introduce twap migrations from: https://github.com/osmosis-labs/osmosis/pull/3729
- test more thoroughly and document

## Testing and Verifying

This change added e2e and migrations tests.

## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? no
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`? no
  - How is the feature or change documented? not applicable

### CODE REVIEW

File: x/twap/migrate.go [hunk 1]
```diff
@@ -14,3 +32,46 @@ func (k Keeper) MigrateExistingPools(ctx sdk.Context, latestPoolId uint64) error
...
+	}
+
+	if len(historicalTimeIndexed) == 0 {
+		return errors.New("error: no historical twap records found")
+	}
+
+	// Since we are iterate over time-indexed records in ascending order,
+	// most recent record should also be updated correctly.
+	for i, record := range historicalTimeIndexed {
+		record := record
```
Comment (mattverse) [inline x/twap/migrate.go:58]: why is this line needed?

File: x/twap/migrate_test.go [hunk 2]
```diff
@@ -93,3 +93,189 @@ func (suite *TestSuite) TestTwapRecord_GeometricTwap_MarshalUnmarshal() {
 	suite.Require().False(originalRecord.GeometricTwapAccumulator.IsNil())
 	suite.Require().Equal(sdk.ZeroDec(), originalRecord.GeometricTwapAccumulator)
 }
+
+func (suite *TestSuite) TestInitializeGeometricTwap() {
+	var (
+		expectedGeometricAccumUpdate = sdk.OneDec()
```
Comment (mattverse) [inline x/twap/migrate_test.go:99]: nit suggestion please feel free to resolve this / disregard this
```suggestion
		defaultGeometricAccumUpdateVal = sdk.OneDec()
```

File: x/twap/migrate_test.go [hunk 3]
```diff
@@ -93,3 +93,189 @@ func (suite *TestSuite) TestTwapRecord_GeometricTwap_MarshalUnmarshal() {
...
+				threeAssetRecordWithPoolIdAndTime(basePoolId, baseTime),
+				threeAssetRecordWithPoolIdAndTime(basePoolId+1, baseTime.Add(time.Second))...),
+
+			expectedMostRecent: map[uint64][]types.TwapRecord{
+				baseRecord.PoolId:     updateGeometricAccum(threeAssetRecordWithPoolIdAndTime(basePoolId, baseTime)),
+				baseRecord.PoolId + 1: updateGeometricAccum(threeAssetRecordWithPoolIdAndTime(basePoolId+1, baseTime.Add(time.Second))),
+			},
+		},
+		"three pools, 2 at the same time, 1 is not, entries and different time for each": {
+			preExistingRecords: merge([][]types.TwapRecord{
```
Comment (mattverse) [inline x/twap/migrate_test.go:183]: why not put in []types.TwapRecord directly instead of providing here [][] and using merge function?

File: x/twap/migrate_test.go [hunk 4]
```diff
@@ -93,3 +93,189 @@ func (suite *TestSuite) TestTwapRecord_GeometricTwap_MarshalUnmarshal() {
...
+			suite.preSetRecords(tc.preExistingRecords)
+
+			// Grab all historical pool indexed twap records prior to initializing the
+			// geometric accumulator.
+			oldHistoricalPoolIndexed, err := k.GetAllHistoricalPoolIndexedTWAPs(ctx)
+			suite.Require().NoError(err)
+			// Initialize the geometric accumulator to get the expected values.
+			expectedHistoricalPoolIndexed := updateGeometricAccum(oldHistoricalPoolIndexed)
+
+			// Repeate for historical time indexed records.
```
Comment (mattverse) [inline x/twap/migrate_test.go:244]:
```suggestion
			// Repeat for historical time indexed records.
```

File: x/twap/migrate.go [hunk 5]
```diff
@@ -14,3 +32,46 @@ func (k Keeper) MigrateExistingPools(ctx sdk.Context, latestPoolId uint64) error
...
+		// Sanity check order.
+		if i > 0 {
+			previousRecord := historicalTimeIndexed[i-1]
+
+			isInvalidOrder := previousRecord.Time.After(record.Time)
+			if isInvalidOrder {
+				return fmt.Errorf("error: historical twap records are not in ascending order, (%v), was after (%v)", previousRecord, record)
+			}
+		}
```
Comment (mattverse) [inline x/twap/migrate.go:67]: Do we have proof that `getAllHistoricalTimeIndexedTWAPs` returns in correctly in ascending order? Looked into existing tests and could not find any relevent test

File: x/twap/migrate.go [hunk 6]
```diff
@@ -14,3 +32,46 @@ func (k Keeper) MigrateExistingPools(ctx sdk.Context, latestPoolId uint64) error
...
+
+// initializeGeometricTwapAcc initializes all existing twap records in state
+// with the given value. Returns nil on success, error otherwise.
+// Returns error if:
+// - fails to get historical time indexed records.
+// - does not find any time indexed records.
+// - the time indexed records returned from state are written back
+// in incorrect order. The order must be ascending by time so that
+// we update the "most recent" index with the right records correctly.
+func (k Keeper) initializeGeometricTwapAcc(ctx sdk.Context, value sdk.Dec) error {
```
Comment (mattverse) [inline x/twap/migrate.go:44]: I think the current tests test this method indirectly by manually calling `updateGeometricAccum`, I think it would be better if we can test this method directly. Wdyt?

**Review by mattverse — COMMENTED** [review]
Main logic looking good, left some questions and concerns about testing. Please take a look!

File: x/twap/migrate.go [hunk 7]
```diff
@@ -1,9 +1,27 @@
...
+
+func NewMigrator(keeper Keeper) Migrator {
+	return Migrator{keeper: keeper}
+}
+
+// Migrate1To2GeometricTwapAcc migrates all existing twap records
+// by initializing their geometric twap accumulator to zero.
+// Return nil on success, error otherwise.
+func (m Migrator) Migrate1To2GeometricTwapAcc(ctx sdk.Context) error {
+	return m.keeper.initializeGeometricTwapAcc(ctx, sdk.ZeroDec())
```
Comment (nicolaslara) [inline x/twap/migrate.go:22]: Shouldn't this be `sdk.OneDec()`? (the identity element of multiplication)

File: x/twap/migrate.go [hunk 8]
```diff
@@ -14,3 +32,46 @@ func (k Keeper) MigrateExistingPools(ctx sdk.Context, latestPoolId uint64) error
...
+		if i > 0 {
+			previousRecord := historicalTimeIndexed[i-1]
+
+			isInvalidOrder := previousRecord.Time.After(record.Time)
+			if isInvalidOrder {
+				return fmt.Errorf("error: historical twap records are not in ascending order, (%v), was after (%v)", previousRecord, record)
+			}
+		}
+
+		record.GeometricTwapAccumulator = value
```
Comment (nicolaslara) [inline x/twap/migrate.go:69]: Can/should we check if `record.GeometricTwapAccumulator` already has a value and skip it if it does?

I'm thinking of a future minor version update from state export (like the issue we just had on 13.1.2) that may try to run this and would end up resetting all geometric accums

File: x/twap/migrate.go (hunk 7 above)
Comment (p0mvn) [reply x/twap/migrate.go:22]: No, it should be zero because we are computing the geometric time-weighted average of prices as an arithmetic time-weighted average of prices in log scale.

File: x/twap/migrate.go (hunk 8 above)
Comment (p0mvn) [reply x/twap/migrate.go:69]: This logic is meant to be run only once. I don't see why and how this would be re-run.

I would be supportive of adding this check. However, it is non-trivial for tests.

The reason is that we are already on v14.x code here. If the proto field exists, then it is initialized to zero.
We cannot set up the test in such a way where we have a v13.x (old) twap record so that its geometric twap accumulator is uninitialized.

File: x/twap/migrate_test.go (hunk 3 above)
Comment (p0mvn) [reply x/twap/migrate_test.go:183]: Well, each call to `threeAssetRecordWithPoolIdAndTime` returns three different []types.TwapRecord that must be merged into a single slice. What is the alternative? We could used `append` multiple times but it's not as clean

File: x/twap/twapmodule/module.go [hunk 9]
```diff
@@ -79,6 +79,12 @@ type AppModule struct {
 
 func (am AppModule) RegisterServices(cfg module.Configurator) {
 	queryproto.RegisterQueryServer(cfg.QueryServer(), grpc.Querier{Q: twapclient.Querier{K: am.k}})
+
+	m := twap.NewMigrator(am.k)
+	err := cfg.RegisterMigration(types.ModuleName, 1, m.Migrate1To2GeometricTwapAcc)
+	if err != nil {
+		panic(err)
+	}
```
[the remark this answers is not in the document]
Comment (p0mvn) [reply x/twap/twapmodule/module.go:87]: What is the problem with doing the registration and increasing the consensus version?

I'm happy to change if there is a problem that I'm unaware of but, AFAIK, the approaches are identical

File: x/twap/migrate.go [hunk 10]
```diff
@@ -14,3 +32,46 @@ func (k Keeper) MigrateExistingPools(ctx sdk.Context, latestPoolId uint64) error
...
+func (k Keeper) initializeGeometricTwapAcc(ctx sdk.Context, value sdk.Dec) error {
+	// In ascending order by time.
+	historicalTimeIndexed, err := k.getAllHistoricalTimeIndexedTWAPs(ctx)
+	if err != nil {
+		return err
+	}
+
+	if len(historicalTimeIndexed) == 0 {
+		return errors.New("error: no historical twap records found")
+	}
```
Comment (czarcas7ic) [inline x/twap/migrate.go:53]: Is it appropriate for this to return an error? It seems like if no historical twap records are found, no logic needs to be run so it should just return nil

File: x/twap/types/keys.go [hunk 11]
```diff
@@ -35,7 +35,7 @@ var (
 
 	// format is just pool id | denom1 | denom2
 	// made for getting most recent key
-	mostRecentTWAPsPrefix = mostRecentTWAPsNoSeparator + KeySeparator
+	MostRecentTWAPsPrefix = mostRecentTWAPsNoSeparator + KeySeparator
```
Comment (czarcas7ic) [inline x/twap/types/keys.go:38]: why do we export this now?

**Comment by p0mvn** [comment]
Closing in favor of: https://github.com/osmosis-labs/osmosis/pull/3972

File: x/twap/migrate.go (hunk 8 above)
Comment (nicolaslara) [reply x/twap/migrate.go:69]: This makes sense. I think we should make sure that it doesn't get run again on an upgrade from export. I'm not sure if this is what would happen though. Just extrapolating from the issue with `initGenesis` on 13.1.2.
